pxr/usd/usdSemantics/labelsQuery.cpp

```diff
@@ -0,0 +1,232 @@
+                           [](const auto t) {
+                                return t >= UsdTimeCode::EarliestTime(); }))
+                           ) {
+                return {};
+            }
+            // Ensure that the queryTime.GetMin() is considered if finite.
+            // If it's not, add the earliest time sample.
+            const double earliest =
+                queryTime.IsMinFinite() ? queryTime.GetMin() :
+                UsdTimeCode::EarliestTime().GetValue();
+            if (times.empty() || (times.front() != earliest)) {
+                times.push_back(earliest);
+            }
+
+            // Times is guaranteed to have interval.GetMin() or
+            // UsdTimeCode::EarliestTime.  If no time samples are authored,
+            // querying at this time sample will return default or fallback
+            // values. No special handling required.
+            TF_DEV_AXIOM(!times.empty());
+
+            _UnorderedTokenSet uniqueLabels;
+            for (const auto timeInInterval : times) {
+                VtTokenArray labelsAtTime;
+                if (!labelsAttr.Get(&labelsAtTime, timeInInterval)) {
+                    TF_RUNTIME_ERROR("Failed to read value at %s",
+                                     UsdDescribe(labelsAttr).c_str());
+                    return {};
+                }
+                uniqueLabels.insert(
+                    std::cbegin(labelsAtTime), std::cend(labelsAtTime));
+            }
+            return uniqueLabels;
+        }
+    }, _time);
+
+    std::unique_lock lock{_cachedLabelsMutex};
+    // If another thread has already computed the cached labels, discard the
+    // results.
+    if (!doubleCheckedLockCondition()) {
+        _cachedLabels.emplace(prim.GetPath(), std::move(labels));
+    }
+    return true;
+}
+
+bool
+UsdSemanticsLabelsQuery::_PopulateInheritedLabels(const UsdPrim& prim) {
+    const auto stage = prim.GetStage();
+    bool hasInheritedLabel = false;
+    for (const auto& path : prim.GetPath().GetAncestorsRange()) {
+        // Note that _PopulateLabels must run every ancestor to update the
+        // cache. Attempting to collapse this expression using |= or the
+        // STL may result in some population being skipped and incorrect
+        // results.
+        if (_PopulateLabels(stage->GetPrimAtPath(path)))
+            hasInheritedLabel = true;
+    }
+    return hasInheritedLabel;
+}
+
+VtTokenArray
+UsdSemanticsLabelsQuery::ComputeUniqueDirectLabels(const UsdPrim& prim)
+{
+    // If the prim was not labeled, we can early exit without
+    // locking.
+    if (!_PopulateLabels(prim)) {
+        return {};
+    }
+
+    std::shared_lock lock{_cachedLabelsMutex};
+    const auto it = _cachedLabels.find(prim.GetPath());
+    if (it == std::cend(_cachedLabels)) {
+        return {};
+    }
+    VtTokenArray result{std::cbegin(it->second), std::cend(it->second)};
+    std::sort(result.begin(), result.end());
```

Comment:
Could we sort as part of population?

Ah, I guess we can't because we use sets so that we can combine inherited results. It seems like it would be nice to frontload the sorting, somehow, but there are tradeoffs to consider, and we'd want to actually measure performance to understand the impact, most likely.

Reply:
It's possible it would be more efficient to use `std::merge` on sorted token arrays.  But perhaps we could address later?